### DISCUSSION TITLE
How to kill java objects from R

### DISCUSSION THREAD

**Discussion by sgum** [s-u/rJava#308] [Q&A] [answered]
Could you advice how can I remove the java-objects from R-code or change .jaddClassPath dynamically? 

I have create the j object by the next script

> model <- "1" 
> 
> jinit()
> .jclassPath()
> .jaddClassPath(str_glue("path/{model}/classes/"))
> .jclassPath()
> quotes.model <- .jnew(str_glue("Model{model}"))
> .jconstructors(quotes.model)

after that script I use quotes.model to make some calculations in R

But, later I need to create new object, but I cant't

> model <- "2"
> 
> jinit()
> .jclassPath()
> .jaddClassPath(str_glue("path/{model}/classes/"))
> .jclassPath()
> quotes.model <- .jnew(str_glue("Model{model}"))
> .jconstructors(quotes.model)

Maybe I need to remove the quotes.model, but I can't

> rm(quotes.model)
Warning message:
In rm(quotes.model) : object 'quotes.model' not found

> class(quotes.model)
[1] "jobjRef"
attr(,"package")
[1] "rJava"

> ls(quotes.model)
Error in as.environment(pos) : S4 object does not extend class "environment"

How I can remove Java-object?

**Comment by s-u** [comment, marked as answer]
You cannot remove loaded classes from Java*.

However what you seem to want is to simply to load another version of the class which you can if you simply use another class loader. That's exactly what `.jpackage(.., own.loader=TRUE)` does so you can have a look here: https://github.com/s-u/rJava/blob/6d362c1e804d85f699e61742e103df0b8634c489/R/loader.R#L42-L50

Once you have a new loader, just use it as `class.loader=` in API calls like `.jnew()`. That way you can load different versions of the classes with the same name many times.

(*) - in theory it's possible if the class loader and all classes that it loaded get garbage-collected in Java. However that is really hard to guarantee since a single reference anywhere in Java is enough to stop that.

**Reply by sgum** [reply]
Thank you for the option, but it looks like we change inside java
